### ISSUE TITLE
Conan 1.x - build-order command is not giving correct output

### ISSUE THREAD

**Issue by mkjkec2005** [conan-io/conan#13143]
### What is your question?

Hi, 

We have 2 components Logger and Basesytem - Logger is dependent on BasySystem. Now we are trying to set up the product pipeline using the example provided in the link below. 

https://github.com/conan-ci-cd-training/products/blob/master/Jenkinsfile

Logger is missing in the build-order, so how can we build the product (Logger) for the change in dependency (BaseSystem/[6.1.0.1])?

+ cat x86_64.lock
{
"graph_lock": {
  "nodes": {
   "1": {
    "ref": "logger/[6.1.0.0#8bcc3a686d18376f5c7e49fb26b6e7b9](https://nam11.safelinks.protection.outlook.com/?url=http%3A%2F%2F6.1.0.0%2F%238bcc3a686d18376f5c7e49fb26b6e7b9&data=05%7C01%7Cmkrishan%40infinera.com%7C4a2dbe56402144a45e6208db0ffb2f5a%7C285643de5f5b4b03a1530ae2dc8aaf77%7C1%7C0%7C638121343752227450%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pa49cBSIwH2MQzjGMi56CgZskydY4gm9yTv12Wc9dKE%3D&reserved=0)",
    "options": "shared=False\nBaseSystem:shared=False",
    "package_id": "991f626f47791da066b00357470672026006e218",
    "prev": "05fafaae1dc866cd9b5b42e7330803a4",
    "requires": [
     "2"
    ],
    "context": "host"
   },
   "2": {
    "ref": "BaseSystem/[6.1.0.1#e69835d7aa1e0df11b949f4e714d550f](https://nam11.safelinks.protection.outlook.com/?url=http%3A%2F%2F6.1.0.1%2F%23e69835d7aa1e0df11b949f4e714d550f&data=05%7C01%7Cmkrishan%40infinera.com%7C4a2dbe56402144a45e6208db0ffb2f5a%7C285643de5f5b4b03a1530ae2dc8aaf77%7C1%7C0%7C638121343752227450%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=OtMCJWjmNV%2BsGeN17mzpKCOI6Gt2SK7DezyoOh37XKs%3D&reserved=0)",
    "options": "shared=False",
    "package_id": "0e10188faca2665db460fb4fbc3832f12077f9b8",
    "context": "host"
   }
  },
  "revisions_enabled": true
},
"version": "0.4",
"profile_host": "[settings]\narch=x86_64\ncompiler=gcc\ncompiler.cppstd=14\ncompiler.libcxx=libstdc++11\ncompiler.version=8\nos=Linux\n[options]\n[build_requires]\n[env]\n"
}
[Pipeline] sh
+ conan lock build-order x86_64.lock --json=build_order.json
[[('[BaseSystem/6.1.0.1@#e69835d7aa1e0df11b949f4e714d550f](mailto:BaseSystem/6.1.0.1@#e69835d7aa1e0df11b949f4e714d550f)', '0e10188faca2665db460fb4fbc3832f12077f9b8', 'host', '2')]]
+ cat build_order.json
[
[
  [
   BaseSystem/6.1.0.1@#e69835d7aa1e0df11b949f4e714d550f,
   "0e10188faca2665db460fb4fbc3832f12077f9b8",
   "host",
   "2"
  ]
]
]
[basesytem_recipe.txt](https://github.com/conan-io/conan/files/10755921/basesytem_recipe.txt)
[Jenkinsfile.txt](https://github.com/conan-io/conan/files/10755922/Jenkinsfile.txt)
[logging_recipe.txt](https://github.com/conan-io/conan/files/10755923/logging_recipe.txt)

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @mkjkec2005 

Maybe you are missing the ``--build=*`` flags? When you are capturing lockfiles and computing "build-order", whatever needs to be built is important. It will not give things that doesn't need to be built. If you want to build it, then using the ``--build`` to specify it is necessary.

A different story is why you want to build it, if you already have a binary. This shouldn't be necessary ever. If you have revisions enabled, that means that the source uniqueness is captured by the recipe revision, the binary configuration is captured by the ``package_id``, if you have already a binary for that ``recipe-revision + package_id`` combination, you don't want to build it again from sources, because it will give you the same binary, but being a waste of CPU, storage, etc.

**Comment by mkjkec2005** [comment]
In the attached Jenkins file i am not very clear how the product build graph is calculated? Is there anything wrong in the below sequence of commands in Jenkinsfile?

sh "conan download ${params.reference} -r ${conan_tmp_repo} --recipe"
sh "conan lock create --reference ${product} --profile=${profile} --lockfile-out=${lockfile} -r ${conan_develop_repo} "

// check if the build-order is empty, this product may not be affected by the changes
// or maybe we don't have to build anything if we are relaunching the builds
sh "conan lock build-order ${lockfile} --json=${bo_file} && cat ${bo_file}"

----------------------------------------------------------------------------------------
Below command reports an error as it is not able to locate the reference product. 

+ conan lock create --reference logger/6.1.0.0@ --profile=x86_64 --lockfile-out=x86_64.lock -r conan-develop
Auto detecting your dev setup to initialize the default profile (/home/bangbuild/external-repo-jenkins/workspace/poc_pip/x86_64/conan_home/.conan/profiles/default)
Found gcc 8
gcc>=5, using the major as version

************************* WARNING: GCC OLD ABI COMPATIBILITY ***********************
 
Conan detected a GCC version > 5 but has adjusted the 'compiler.libcxx' setting to
'libstdc++' for backwards compatibility.
Your compiler is likely using the new CXX11 ABI by default (libstdc++11).

If you want Conan to use the new ABI for the default profile, run:

    $ conan profile update settings.compiler.libcxx=libstdc++11 default

Or edit '/home/bangbuild/external-repo-jenkins/workspace/poc_pip/x86_64/conan_home/.conan/profiles/default' and set compiler.libcxx=libstdc++11

************************************************************************************



Default settings
	os=Linux
	os_build=Linux
	arch=x86_64
	arch_build=x86_64
	compiler=gcc
	compiler.version=8
	compiler.libcxx=libstdc++
	build_type=Release
*** You can change them in /home/bangbuild/external-repo-jenkins/workspace/poc_pip/x86_64/conan_home/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***


logger/6.1.0.0: Retrieving from server 'conan-develop' 
logger/6.1.0.0: Trying with 'conan-develop'...
ERROR: logger/6.1.0.0 was not found in remote 'conan-develop'

**Comment by memsharded** [comment]
This doesn't seem a build-order issue, just that the ``logger/6.1.0.0`` is not in the server repo, isn't it?

**Comment by mkjkec2005** [comment]
Yes, exactly. In this case product pipeline should build the product
(logger/6.1.0.0). Lockfile contains 2 nodes - logger and basesystem but
build-order returns only one.
May be I am missing some important concept here. Kindly help.

**Comment by memsharded** [comment]
I think this is what I commented above. When you are capturing the lockfile you need to pass ``--build=*``, otherwise the lockfile captures an existing binary, and it doesn't need to be built, so "build-order" doesn't return it.
And please also consider the other observation: if a binary for it already exists, why do you want to rebuild it?

**Comment by mkjkec2005** [comment]
Build flag (--build=missing) was not working during lock creation. Now I
have deleted Logger binary from Artifactory, and my expectation is that the
product pipeline will build logger as there is a change in
basesystem(product pipeline is triggered with basesystem recipe reference).

**Comment by memsharded** [comment]
I am afraid this is not how it works.
A lockfile is just a snapshot of a dependency graph in a given time. If it captures a snapshot with a given package, that exists, it has an existing binary that doesn't have to be built, removing such binary from Artifactory will not cause the lockfile to be aware of that and rebuild it. It is actually the opposite, the snapshot is there to guarantee that that exact binary will be used, and not other, not even rebuild it.

It is very important, if you want to build it, to capture the lockfile with ``--build=<what you want to build>``. If it is not working, please report it.

Also, I'd strongly encourage to not implement that yet in Jenkins. Build first a proof of concept locally, with simple python scripts in your machine, and only move to jenkins when everything is very clear. Developing this kind of task in Jenkins directly is simply a huge pain.

**Comment by mkjkec2005** [comment]
Understood. But, If one of the dependency captured in Lockfile (assume that
the snapshot it captured is week old) is deleted ( may be by a cron job
which cleans up one week old packages from Artifactory), what is the next
course of action? Capture rhe lockfile with --build=missing flag?

**Comment by memsharded** [comment]
Yes, definitely ``--build=missing`` is the way to go in general.
I was explaining ``--build=*`` before, just to explain the need of the flag, but building all is extremely rarely needed, and shouldn't be done in CI. 

The good approach is asking: "what needs to be built?" with the ``--build=missing`` and rely on the ``package_id`` computation (in 1.X you might want to check the defaults package_id_modes, to make them more strict, in Conan 2.0 the defaults are safer).

**Comment by mkjkec2005** [comment]
So, the issue here is that there are some changes in the command in the example provided in CI CD reference material:

**Example provided in CI CD reference Jenkinsfile:**    
	stage("Check if the new revision  is in product graph") {

                // install just the recipe for the new revision of the lib and then resolve graph with develop repo
                // we don't want to bring binaries at this moment, this could be just a agent that
                // then sends lockfiles to other nodes to build them
                sh "conan download ${params.reference} -r ${conan_tmp_repo} --recipe"
                sh "conan graph lock ${product} --profile=${profile} --lockfile=${lockfile} -r ${conan_develop_repo}"

                // check if the build-order is empty, this product may not be affected by the changes
                // or maybe we don't have to build anything if we are relaunching the builds
                sh "conan graph build-order ${lockfile} --json=${bo_file} --build missing"
                build_order = readJSON(file: bo_file)
                if (build_order.size()>0) {
                  affected_product = true
                }
              }

**Changes in Jenkinsfile to make it work**

  stage("Check if the new revision  is in product graph") {

                // install just the recipe for the new revision of the lib and then resolve graph with develop repo
                // we don't want to bring binaries at this moment, this could be just a agent that
                // then sends lockfiles to other nodes to build them
                sh "conan download ${params.reference} -r ${conan_tmp_repo} --recipe"
                sh "conan lock create --reference ${product} --profile=${profile} --lockfile-out=${lockfile} -r ${conan_develop_repo} **--build** && cat ${lockfile}"

                // check if the build-order is empty, this product may not be affected by the changes
                // or maybe we don't have to build anything if we are relaunching the builds
                sh "conan lock build-order ${lockfile} --json=${bo_file} && cat ${bo_file}"
                build_order = readJSON(file: bo_file)
                if (build_order.size()>0) {
                  affected_product = true
                }
              }

**Comment by memsharded** [comment]
Yes, please note that the ci_cd_training repo is not advertised or linked from our official docs in https://docs.conan.io

This is because this material is not updated nor maintained. It shouldn't be used as a good reference. It might contain some helpful bits, but it is not reference material.

**Comment by mkjkec2005** [comment]
Thanks. When can we expect this training for Conan 2.0?

**Comment by memsharded** [comment]
There will be docs, most likely in tutorial format with accompanying examples code repo, in the following months, I'd expect something in the next 2-3 months.

**Comment by mkjkec2005** [comment]
Thank you. Another question just for my understanding - how to generate
recipe revision ? Does conan use some python module to calculate the hash?

**Comment by memsharded** [comment]
The recipe revision is the hash of the contents of the recipe. Conan does this internally.
Is it possible to use the git commit instead with ``revision_mode = "scm"`` in the recipe. In both cases, they work because the revision guarantees uniqueness of the contents (not possible to change contents without changed revision, and viceversa).

**Comment by mkjkec2005** [comment]
Thank you. I am trying to play with the lockfile and build order. In the below command, how can I ensure that only the fileutils and DbCrypto are built?

**Note:** 

1. Dbcrypto depends on Fileutils and Logger
2. Logger depends on Basesystem
3. Input to the product pipeline is a recipe reference for FileUtils (FileUtils/6.1.0.0@#bc12fb1eadb36be5346214137b575831) for a change in associated PR-1.
4. I have the Logger and base system binaries already available in the conan-develop repo. 

Below is the command used:
conan lock create --reference dbcrypto/6.1.0.1@ --profile=x86_64 --lockfile-out=x86_64.lock -r conan-develop --build

I don't want to rebuild the base system and logger, but it is building everything. **The log is attached for your reference - lock_query.txt.** 

BaseSystem/6.1.0.1: Forced build from source
FileUtils/6.1.0.0: Forced build from source
logger/6.1.0.0: Forced build from source
dbcrypto/6.1.0.1: Forced build from source
Requirements
    BaseSystem/6.1.0.1 from 'conan-develop' - Downloaded
    FileUtils/6.1.0.0 from 'conan-tmp' - Cache
    dbcrypto/6.1.0.1 from 'conan-develop' - Downloaded
    logger/6.1.0.0 from 'conan-develop' - Downloaded
Packages
    BaseSystem/6.1.0.1:0e10188faca2665db460fb4fbc3832f12077f9b8 - Build
    FileUtils/6.1.0.0:0e10188faca2665db460fb4fbc3832f12077f9b8 - Build
    dbcrypto/6.1.0.1:735bda2ed61fb7aa2754d2330e07ffcdfde0d401 - Build
    logger/6.1.0.0:b3ea592a24d3a59021fecdd1c1cb9d22219b98bd - Build

**Comment by mkjkec2005** [comment]
[lock_query.txt](https://github.com/conan-io/conan/files/10789231/lock_query.txt)

**Comment by memsharded** [comment]
In general, you don't need to tell what to build, and ``--build=missing`` is what will automatically compute what needs to be built from sources and what not. It will use the different ``package_id_mode``, analyze what versions, revisions and binaries changed in the graph, and propagate those changes to the consumers to compute if they need to be rebuilt. 
If it is telling you that you need to build more packages it is because they should be rebuilt.

If you used ``--build=*``, then you need to relax it, and say ``--build=missing``, and then you will not need to build all packages.


More hints: Do not use uppercase in your package names, use always lowercase, as uppercase will be restricted in conan 2.0

**Comment by mkjkec2005** [comment]
What is the best way to automatically rebuild downstream components if a public header is changed for an upstream component (rebuild a package when there are changes to the public headers of a dependency)?

**Comment by memsharded** [comment]
> What is the best way to automatically rebuild downstream components if a public header is changed for an upstream component (rebuild a package when there are changes to the public headers of a dependency)?

There is no single "best" approach for everyone, because it depends on many factors.
But something in this line makes sense, which is part of what 2.0 proposes:

- Use ``minor_mode``. That means, you control if your changes need to rebuild consumers or not:
   - Changing public headers to add new things, but don't touch existing => you can bump the patch only safely
   - Changing public headers, in a non breaking way, but requiring building consumers => bump the minor
   - Breaking changes to public headers => bump the major

The ``minor_mode`` will require the rebuild of consumers when the minor is bumped, but not if the patch is bumped.

You do this, and ``--build=missing`` will take care to compute what needs to be rebuilt in the whole dependency graph, automatically and optimally.

**Comment by mkjkec2005** [comment]
What is the equivalent in Conan 1.x? I am currently running package and
product pipeline in context of 1.x. Once some basic use cases are showcased
with 1.x, we can switch to 2.x. Our management is keen to get the POC
working with 1.x.

As per my understanding, we need to use recipe_revision or package_version
(package ID mode) to get this working.

**Comment by memsharded** [comment]
> What is the equivalent in Conan 1.x? I am currently running package and
product pipeline in context of 1.x. Once some basic use cases are showcased
with 1.x, we can switch to 2.x. Our management is keen to get the POC
working with 1.x.

Yes, the above is implementable also in 1.X. There is ``default_package_id_mode`` you can define in ``conan.conf``
It can also define several modes, like ``minor_mode`` and ``recipe_revision_mode``. Which one, really depends on your needs, if you have good discipline bumping versions, ``minor_mode`` will be more efficient, if not ``recipe_revision_mode`` is safer.


> Our management is keen to get the POC working with 1.x.

Sorry to hear that. This indeed sounds like quite some unnecessary pain.

**Comment by mkjkec2005** [comment]
Our experiments with 1.x are going well. I have another query on use case#2 below:
1)	Header file which has source code/ function definition(without any interface change)
2)	Header file has an interface change which would need to change in the downstream module as well.

Taking the earlier example of DbCrypto product:

- Dbcrypto depends on Fileutils and Logger
- Logger depends on Basesystem

If we make an interface change in Basesystem, which needs a change in logger also, how do we validate (or build) these changes together in CI using Conan? (package pipelines are separate, so base system and logger will build independently and fail compilation).

**Comment by memsharded** [comment]
If a change in Basesystem requires changing the code of the consumers, by definition it is a breaking change. The process would be:
- If the used ``basesystem`` version is ``1.3`` (note using always lowercase package names, that will be mandatory in 2.0), the breaking change would create ``basesystem/2.0`` major version. We upload the version to the "builds" server repository
- ``logger/2.1`` contains a ``requires = "basesystem/[>=1.0 <2.0]``, so when we check if it builds, it keeps using previous ``basesystem/1.3``, nothing changed, nothing needs to be rebuilt, done. Same with ``dbcrypto``, nothing to do, ready.
- The "downstream products" pipeline is green, nothing broken because the new 2.0 major version is not pulled. We are done, we can merge the PR and promote/copy the "basesystem/2.0" binaries from the server "build" repo to the server "develop" repo, so we make it available for others
- Now the ``logger/2.1`` team start doing the changes in the code of logger, starting by bumping the requirement to ``requires = "basesystem/[>=2.0 <3.0]"``. Then they modify the code of logger until it locally works. Then, depending on the necessary code changes, they decide what version logger bumps. If all necessary code changes are in .cpp files, it is possible to not update the version, because consumers of logger will not see any breaking thing in the logger interfaces. Lets say that we still decide to release a new minor ``logger/2.2``. We do the PR, ``logger/2.2`` is created and uploaded to "build" server repo
- The products pipeline fire the build of ``dbcrypto`` that will automatically pull ``logger/2.2`` because it is in its valid version range. It can build a new binary for ``dbcrypto`` that is also uploaded to the "build" server repo.
- If everything is good and green, the PR for ``logger/2.2`` can be merged, and the packages for logger and dbcrypto copied from the "build" server to the "develop" server.

**Comment by mkjkec2005** [comment]
Thank you for the details. The issue is that we don't have good discipline in maintaining and bumping versions. So i am trying the experiments with package ID mode set to recipe_revision_mode.

**Comment by memsharded** [comment]
Ok, still same principle applies, the same flows will happen, just there will be more builds from sources, as whenever a dependency changes it will require its consumer to be re-built from sources.

**Comment by mkjkec2005** [comment]
yes, but ultimately we need to deal with versions only to mark a breaking change ( breaking change would create basesystem/2.0 major version) and then update the version range in Logger (can't be done at the recipe revision level). Correct?
Also, is this the same way it is handled in 2.x?
What if both basesystem and logger are interdependent - both need to be compiled together to get the build working? locally can be solved using editables bu not sure how that can be handled in CI?

**Comment by memsharded** [comment]
> yes, but ultimately we need to deal with versions only to mark a breaking change ( breaking change would create basesystem/2.0 major version) and then update the version range in Logger (can't be done at the recipe revision level). Correct?

Versions are important not only for breaking changes, they also encapsulate changelogs, history, needs of rebuilding consumers, being ready for new compiler versions or standards... There are many reasons you want also to do minor or patches releases.

Not sure what you mean that it can't be done at the recipe revision level.

> Also, is this the same way it is handled in 2.x?

Yes, none of this is specific to Conan really. This is more or less a common thing in package management, and software versioning in general.

> What if both basesystem and logger are interdependent 

If they are really that dependent, they are not separate packages they should live in the same package. If you cannot build, release and reuse each one separately, then they are not separate packages.

In reality, it is typically good to do incremental changes, you bump the version of the baseystem, without being included in the downstream range, build it, then send the PR of the logger updating the ``requires`` to include the ``basesystem`` new version

**Comment by mkjkec2005** [comment]
Something is not correctly set up with minor_mode. I have the following package id defined in conan.conf (globally). 

default_package_id_mode = minor_mode

Now I am making 2 changes in Basesytem - changing one header file and bumping version in version.txt from 6.1.0.0 to 6.1.0.1. 

Logger conanfile includes a requires to basesystem
    def requirements(self):
        self.requires("baseyystem/[>=6.1.0 <6.2.0]")

PR opened in basesytem triggers package pipeine and subsequently product pipeline. In product pipeline, my understanding is that it should not rebuild logger ( as only build version changed). But it is building it from sources. 

What could be the issue?

**Comment by memsharded** [comment]
Hi @mkjkec2005 

Sorry this was not followed up back then.

Now it is most likely that this ticket is outdated, so I am closing it. Please don't hesitate to open new tickets for any further question or issue. Thanks for the feedback!

For reference, there is a new section in the Conan 2 docs that describes CI good practices for scale with ``conan graph build-order``, see https://docs.conan.io/2/ci_tutorial/tutorial.html
